### PR TITLE
Make ByteSizeUnit implements Writeable

### PR DESCRIPTION
**Pull request by tlrx**
This commit makes `ByteSizeUnit` implement `Writeable`.

### CODE REVIEW

**Review by javanna — CHANGES_REQUESTED** [review]
left two small comments, LGTM otherwise

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java [hunk 1]
```diff
@@ -225,6 +225,35 @@ public long toPB(long size) {
...
+        this.id = id;
+    }
+
+    public byte getId() {
+        return id;
+    }
+
+    public static ByteSizeUnit fromId(byte id) {
+        switch (id) {
+            case 0:
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java]: Yes

File: core/src/test/java/org/elasticsearch/common/unit/ByteSizeUnitTests.java [hunk 2]
```diff
@@ -77,4 +80,21 @@ public void testPB() {
...
+
+                try (StreamInput in = out.bytes().streamInput()) {
+                    ByteSizeUnit deserialized = in.readByteSizeUnit();
+                    assertEquals(unit, deserialized);
+                }
+            }
+        }
+
+        final byte randomId = (byte) randomIntBetween(6, Byte.MAX_VALUE);
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> ByteSizeUnit.fromId(randomId));
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/common/unit/ByteSizeUnitTests.java]: Ok

**Comment by tlrx** [comment]
@javanna I updated the code according to your comments.

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java [hunk 3]
```diff
@@ -225,6 +234,23 @@ public long toPB(long size) {
 
     static final long MAX = Long.MAX_VALUE;
 
+    private final byte id;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java]: nit: how about simply using the ordinal of the enum? we would have to be careful with the order if we add more units, which we are not likely to do. Also, in any case we would have to be careful with be comp anyways.

**Review by javanna — APPROVED** [review]
left a minor comment, LGTM though

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 4]
```diff
@@ -572,6 +573,13 @@ public GeoPoint readGeoPoint() throws IOException {
     }
 
     /**
+     * Reads a {@link ByteSizeUnit} from this stream input
+     */
+    public ByteSizeUnit readByteSizeUnit() throws IOException {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: I think this should just be a static method on `ByteSizeUnit`.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: We do that in a lot of places in queries and aggregations.

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java [hunk 5]
```diff
@@ -32,11 +32,11 @@
 public class ByteSizeValue implements Writeable {
 
     private final long size;
-    private final ByteSizeUnit sizeUnit;
+    private final ByteSizeUnit unit;
 
     public ByteSizeValue(StreamInput in) throws IOException {
         size = in.readVLong();
-        sizeUnit = ByteSizeUnit.BYTES;
+        unit = ByteSizeUnit.BYTES;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Did you want to do the reading/writing of the unit here? That wouldn't make the PR too huge and it seems like what we want?

**Review by nik9000 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java (hunk 3 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java]: Sure. I usually don't like to use ordinal to serialize enums but I agree, we're not going to add new units anytime soon :)

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java (hunk 5 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: I'm changing this one thing at a time - next will be ByteSizeValue serialization. So if you're OK I'd like to keep this as it is for now and change it in a follow up PR.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Fine be me.

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java [hunk 6]
```diff
@@ -216,6 +222,15 @@ public long toPB(long size) {
         }
     };
 
+    private static final ByteSizeUnit[] UNITS = new ByteSizeUnit[ByteSizeUnit.values().length];
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java]: I think this is just a duplicate of `BytesSizeUnit.values()`?

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java (hunk 5 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Thanks Nik

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java (hunk 6 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeUnit.java]: Raah... yes, I blindly updated the code but this is not needed anymore.

**Comment by tlrx** [comment]
retest this please

**Review by nik9000 — APPROVED** [review]

**Comment by tlrx** [comment]
Thanks @nik9000 @javanna !

**Pull request merged**